### PR TITLE
591 add DecathlonDataset

### LINKED ISSUE #591: medical segmentation decathlon dataset as MONAI dataset
**Is your feature request related to a problem? Please describe.**
this ticket looks for the integration of [the medical segmentation decathlon datasets](https://drive.google.com/drive/folders/1HqEgzS8BV2c7xYNrZdEAnrHk7osJJ--2)
following the pattern in [monai.applications.MedNISTDataset](https://github.com/Project-MONAI/MONAI/blob/master/monai/application/datasets.py#L22)

the data need to be hosted elsewhere to allow for simple/robust automatic downloading

(related to https://github.com/Project-MONAI/MONAI/issues/246)

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #591 .

### Description
This PR implemented `DecathlonDataset` based on our public dataset arch.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

**Comment by Nic-Ma** [comment]
WIP because of missing MD5 values of Task03, Task06, and Task07.
Thanks.

**Comment by Nic-Ma** [comment]
> WIP because of missing MD5 values of Task03, Task06, and Task07.
> Thanks.

Got MD5 of all the data and verified locally.

**Comment by Nic-Ma** [comment]
Hi @wyli , @ericspod  and @atbenmurray ,

1. I already downloaded all the Decathlon data and verified the MD5 values locally.
2. The CI test passed for all platforms, then I added `skip_if_quick` to skip this downloading test.
3. I will develop a tutorial to show how to use `MedNISTDataset` and `DecathlonDataset` and how to add new dataset in a separate PR soon.

Could you please help review it?
Thanks.

File: monai/application/datasets.py [hunk 1]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
...
+
+
+class DecathlonDataset(Randomizable, CacheDataset):
+    """
+    The Dataset to automatically download the data of Decathlon challenge (http://medicaldecathlon.com/)
+    and generate items for training, validation or test. It's based on `CacheDataset` to accelerate
+    the training process.
+
+    Args:
+        root_dir: target directory to download and load Decathlon dataset.
```
Comment (wyli) [inline monai/application/datasets.py:132]: docstring could be: "root_dir: User's local directory for caching and loading the MSD datasets."

File: monai/application/datasets.py [hunk 2]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
...
+        "Task01_BrainTumour": "https://drive.google.com/uc?id=1A2IU8Sgea1h3fYLpYtFb2v7NYdMjvEhU",
+        "Task02_Heart": "https://drive.google.com/uc?id=1wEB2I6S6tQBVEPxir8cA5kFB8gTQadYY",
+        "Task03_Liver": "https://drive.google.com/uc?id=1jyVGUGyxKBXV6_9ivuZapQS8eUJXCIpu",
+        "Task04_Hippocampus": "https://drive.google.com/uc?id=1RzPB1_bqzQhlWvU-YGvZzhx2omcDh38C",
+        "Task05_Prostate": "https://drive.google.com/uc?id=1Ff7c21UksxyT4JfETjaarmuKEjdqe1-a",
+        "Task06_Lung": "https://drive.google.com/uc?id=1I1LR7XjyEZ-VBQ-Xruh31V7xExMjlVvi",
+        "Task07_Pancreas": "https://drive.google.com/uc?id=1YZQFSonulXuagMIfbJkZeTFJ6qEUuUxL",
+        "Task08_HepaticVessel": "https://drive.google.com/uc?id=1qVrpV7vmhIsUxFiH189LmAn0ALbAPrgS",
+        "Task09_Spleen": "https://drive.google.com/uc?id=1jzeNU1EKnK81PyTsrx0ujfNl-t0Jo8uE",
+        "Task10_colon": "https://drive.google.com/uc?id=1m7tMpE9qEcQGQjL_BdMD-Mvgmc44hG1Y",
```
Comment (wyli) [inline monai/application/datasets.py:169]: Task10_colon -> Task10_Colon

File: monai/application/datasets.py [hunk 3]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
                 raise ValueError("section name can only be: training, validation or test.")
             data.append({"image": image_files_list[i], "label": image_class[i]})
         return data
+
+
+class DecathlonDataset(Randomizable, CacheDataset):
+    """
+    The Dataset to automatically download the data of Decathlon challenge (http://medicaldecathlon.com/)
+    and generate items for training, validation or test. It's based on `CacheDataset` to accelerate
```
Comment (wyli) [inline monai/application/datasets.py:128]: docstring: CacheDataset -> :py:class:`monai.data.CacheDataset`

File: monai/application/datasets.py [hunk 4]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
...
+        self.set_random_state(seed=seed)
+        if task not in self.resource:
+            raise ValueError(f"unsupported task: {task}.")
+        dataset_dir = os.path.join(root_dir, task)
+        tarfile_name = f"{dataset_dir}.tar"
+        if download:
+            download_and_extract(self.resource[task], tarfile_name, root_dir, self.md5[task])
+
+        if not os.path.exists(dataset_dir):
+            raise RuntimeError("can not find dataset directory, please use download=True to download it.")
```
Comment (wyli) [inline monai/application/datasets.py:210]: need more info for this error: `f"cannot find the directory {dataset_dir}"`....

File: monai/application/datasets.py [hunk 5]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
...
+        cache_rate: float = 1.0,
+        num_workers: int = 0,
+    ):
+        if not os.path.isdir(root_dir):
+            raise ValueError("root_dir must be a directory.")
+        self.section = section
+        self.val_frac = val_frac
+        self.set_random_state(seed=seed)
+        if task not in self.resource:
+            raise ValueError(f"unsupported task: {task}.")
```
Comment (wyli) [inline monai/application/datasets.py:203]: print supported tasks: `f"available options are {list(self.resource)}."`

File: monai/application/datasets.py [hunk 6]
```diff
@@ -16,6 +16,7 @@
 from typing import Any, Callable
```
Comment (wyli) [inline monai/application/datasets.py:16]: change application -> apps as we discussed

File: monai/application/datasets.py [hunk 7]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
                 raise ValueError("section name can only be: training, validation or test.")
             data.append({"image": image_files_list[i], "label": image_class[i]})
         return data
+
+
+class DecathlonDataset(Randomizable, CacheDataset):
+    """
+    The Dataset to automatically download the data of Decathlon challenge (http://medicaldecathlon.com/)
```
Comment (wyli) [inline monai/application/datasets.py:127]: of Decathlon -> of Medical Segmentation Decathlon

File: monai/application/datasets.py [hunk 8]
```diff
@@ -120,3 +120,114 @@ def _generate_data_list(self, dataset_dir):
...
+        val_frac: percentage of of validation fraction from the `training` section, default is 0.2.
+            Decathlon data only contains `training` section with labels and `test` section without labels,
+            so randomly select fraction from the `training` section as the `validation` section.
+        cache_num: number of items to be cached. Default is `sys.maxsize`.
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        cache_rate: percentage of cached data in total, default is 1.0 (cache all).
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        num_workers: the number of worker threads to use.
+            If 0 a single thread will be used. Default is 0.
+
```
Comment (wyli) [inline monai/application/datasets.py:151]: could put a quick code example here, something like:
```
Examples::

       >>>  DecathlonDataset()
```

File: monai/application/datasets.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:132]: updated in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:169]: updated in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:128]: updated in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:210]: updated both MedNISTDataset and DecathlonDataset in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:203]: updated in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 7 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:127]: updated in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 6 above)
Comment (Nic-Ma) [reply monai/application/datasets.py:16]: updated in the latest commit.
Thanks.

File: monai/application/datasets.py (hunk 8 above)
Comment (Nic-Ma) [reply monai/application/datasets.py]: added example code in the latest commit.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review!
I updated the PR according to your comments.
Could you please help review it again?
Thanks.

**Pull request merged**